sound/soc/sof/pcm.c

```diff
@@ -482,9 +482,6 @@ static int sof_pcm_open(struct snd_soc_component *component,
 	snd_pcm_hw_constraint_step(substream->runtime, 0,
 				   SNDRV_PCM_HW_PARAM_BUFFER_BYTES,
 				   le32_to_cpu(caps->period_size_min));
-	snd_pcm_hw_constraint_step(substream->runtime, 0,
-				   SNDRV_PCM_HW_PARAM_PERIOD_BYTES,
-				   le32_to_cpu(caps->period_size_min));
```

Comment:
with this you remove ALL constraints on the period size.
I would assert the period size needs to be at least caps->period_size_min.

You also need to add in your commit message that choosing such periods will results in non-periodic events reported, and while it may be nice for some applications it may be problematic if the application checks when the period is reported.

Reply:
@plbossart wrote
> I would assert the period size needs to be at least caps->period_size_min

The minimum is set a few lines down in pcm_open:
runtime->hw.period_bytes_min = le32_to_cpu(caps->period_size_min);

But, but, I now wonder should we modify the buffer size constraint as well. This could result in weird period/buffer size combos, if we restrict only one.

> You also need to add in your commit message that choosing such periods will results in non-periodic events reported

I tried to capture this with "While this approach helps user-space to avoid..." sentence. OTOH, if any application is surprised by this, then it will be surprised by any USB audio device as well.

Reply:
yes, I was wondering how you can have a buffer that is required to be a multiple of 1ms while allowing for periods that are not?

Reply:
@plbossart Upon some further investigation, I see now reason to constraint the buffer size either, other than min/max. Patch and commit message updated.

Reply:
@kv2019i don't we want the buffer size to be a multiple of the period size? I can't recall what this constraint is but that seems rather useful to enforce, no?

Reply:
@plbossart Yes, been looking at that as well. That seems like a very reasonable constraint (i.e. any app that uses period-based wake-ups, probably wants an even multiple), but its use in drivers seem somewhat adhoc. I'm not really convinced drivers should add this constraint unless it is something the hw enforces.

For the HDA implementation in SOF, we even have code in hda_dsp_stream_setup_bdl() to explicitly handle the case where buffersize is not a multiple of period size. OTOH, legacy HDA driver does set the integer constraint and there is some code in legacy driver that at least seems suspicious (e.g. snd_hdac_stream_setup_periods(), SOF is not using this). But as a good example, the HDA legacy driver line that added the constraint is from a commit done in 2005, with no rationale why integer periods is used -- just a limitation of current code, or something in the hw.

So adding this at SOF common level does not seem right. There are many widely used drivers that do not set such constraint and applications work fine with them.

Let me stress test this a bit more.

Reply:
@kv2019i maybe it's true for HDAudio but I am not sure this would work for SOF in general, e.g. for BYT/BDW legacy platforms where the DMA is handled by the firmware. To the best of my knowledge those DMAs only handle linear memory fragments, and if you have a period straddling across the end and the start of the ring buffer we may have a problem.